acra-censor/acra-censor_test.go

```diff
@@ -103,68 +107,322 @@ func testWhitelistTables(t *testing.T, acraCensor *AcraCensor, whitelistHandler
+		"SELECT A, B",
+		"SELECT X1, X2",
+		"SELECT col1, col2",
+	}
+	blockableQueries := []string{
+		"SELECT A",
+		"SELECT A, B, C",
+		"SELECT A, B, C, D",
+	}
+	for _, query := range acceptableQueries {
+		err = censor.HandleQuery(query)
+		if err != nil {
+			t.Fatal(err, "Whitelist pattern blocked query. \nPattern:", pattern, "\nQuery:", query)
+		}
+	}
+	for _, query := range blockableQueries {
+		err = censor.HandleQuery(query)
 		if err != handlers.ErrWhitelistPatternMismatch {
-			t.Fatal(err)
+			t.Fatal(err, "Whitelist pattern passed query. \nPattern:", pattern, "\nQuery:", query)
 		}
 	}
-	queryIndexesToPass := []int{0, 4}
-	//acracensor should not block those queries
-	for _, i := range queryIndexesToPass {
-		err := acraCensor.HandleQuery(testQueries[i])
+
+	whitelist.Reset()
+	pattern = "SELECT A, %%COLUMN%% FROM testTable"
+	err = whitelist.AddPatterns([]string{pattern})
+	if err != nil {
+		t.Fatal(err)
+	}
+	acceptableQueries = []string{
+		"SELECT A, B FROM testTable",
+		"SELECT A, X2 FROM testTable",
+		"SELECT A, col2 FROM testTable",
+	}
+	blockableQueries = []string{
+		"SELECT A FROM testTable",
+		"SELECT B, A FROM testTable",
+		"SELECT A, B, C FROM testTable",
+		"SELECT A, B, C, D FROM testTable",
+	}
+	for _, query := range acceptableQueries {
+		err = censor.HandleQuery(query)
 		if err != nil {
-			t.Fatal(err)
+			t.Fatal(err, "Whitelist pattern blocked query. \nPattern:", pattern, "\nQuery:", query)
+		}
+	}
+	for _, query := range blockableQueries {
+		err = censor.HandleQuery(query)
+		if err != handlers.ErrWhitelistPatternMismatch {
+			t.Fatal(err, "Whitelist pattern passed query. \nPattern:", pattern, "\nQuery:", query)
+		}
+	}
+
+	pattern = "SELECT %%COLUMN%%, B FROM testTable"
+	err = whitelist.AddPatterns([]string{pattern})
+	if err != nil {
+		t.Fatal(err)
+	}
+	acceptableQueries = []string{
+		"SELECT A, B FROM testTable",
+		"SELECT X2, B FROM testTable",
+		"SELECT col2, B FROM testTable",
+	}
+	blockableQueries = []string{
+		"SELECT A FROM testTable",
+		"SELECT B, A FROM testTable",
+		"SELECT A, B, C FROM testTable",
+		"SELECT A, B, C, D FROM testTable",
+	}
+	for _, query := range acceptableQueries {
+		err = censor.HandleQuery(query)
+		if err != nil {
+			t.Fatal(err, "Whitelist pattern blocked query. \nPattern:", pattern, "\nQuery:", query)
+		}
+	}
+	for _, query := range blockableQueries {
+		err = censor.HandleQuery(query)
+		if err != handlers.ErrWhitelistPatternMismatch {
+			t.Fatal(err, "Whitelist pattern passed query. \nPattern:", pattern, "\nQuery:", query)
+		}
+	}
+}
+func testWhitelistWherePattern(t *testing.T) {
+	var err error
+
+	censor := NewAcraCensor()
+	defer censor.ReleaseAll()
+	whitelist := handlers.NewWhitelistHandler()
+	censor.AddHandler(whitelist)
+
+	pattern := "SELECT a, b, c FROM z %%WHERE%%"
+	err = whitelist.AddPatterns([]string{pattern})
+	if err != nil {
+		t.Fatal(err)
+	}
+	acceptableQueries := []string{
+		"SELECT a, b, c FROM z WHERE a = 'someValue'",
```

Comment:
can we test here queries that we use for star pattern too (https://github.com/cossacklabs/acra/blob/acracensor/acra-censor/acra-censor_test.go#L545) ? to test different `where` clauses without star and columns